src/librustc_mir/transform/add_call_guards.rs

```diff
@@ -46,8 +46,9 @@ impl AddCallGuards {
 
         let cur_len = body.basic_blocks().len();
 
-        for block in body.basic_blocks_mut() {
-            match block.terminator {
+        for bb in body.basic_blocks().indices() {
+            let is_cleanup = body.basic_blocks()[bb].is_cleanup;
```

Comment:
In a lot of cases were I have to iterator over the indices rather than the elements, is there an alternative so we're not paying the cost of the bounds check each iteration?

Reply:
you can use `iter_enumerated_mut` instead of `indices`, then you get the index and the body directly

Reply:
oh, this change is because I asked you to remove `basic_blocks_mut()`? hmm... maybe we should keep it but add a `and_invalidate_cache` to the name

Reply:
`basic_blocks_mut` still exists, but does not invalidate the cache, I moved that into `Body::basic_block_terminator(_opt)_mut`. Accessing the terminator mutably requires a unique borrow of `body` (since we need to modify `predecessors_cache` within `body`, and we need to return `&mut Terminator` which comes from a `BasicBlockData` that's also within `body`). There's no way to do that if we're returning unique references to the `BasicBlockData` elements, since that will maintain a unique borrow of `body`, preventing us from using it to also call the terminator fn.

Reply:
I think that could work, though I might want to do that as a separate PR since I just did a quick analysis and there are a lot more usages of `statements` than there are of `terminator`. I think I could just move back to having everything invalidate inside of `basic_blocks_mut` for now.

Maybe for the statements something like `basic_blocks_only_statements_mut` (what a horrible name I've come up with :sweat_smile:) that exposes the other items through a normal ref, but you can access statements mutably? Or is it rather we want a version where everything else is still mutable, but terminator cannot be accessed mutably? There's also the `is_cleanup` property to consider.

Reply:
Yea let's do it in a future PR, names aside, I like the statement thing